From 5a8ae372c2ffcf95d0b7b13ecd6d649733d12839 Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Fri, 16 Oct 2026 18:53:48 +0200
Subject: [PATCH] QIcon: Don't modify the dpr of a cached pixmap entry

When QIcon delay-loads a pixmap don't set the dpr to the currently
requested dpr as this modifies the original entry which might result in
returning a wrong pixmap for a requested size afterwards.
It's not needed to set the dpr here because the dpr is set later on the
returned pixmap. Setting a new dpr on a pixmap will detach (which was
the reason to set the dpr in the first place) but since the so modified
pixmaps are stored in the QPixmapCache, no detach will occur anymore.

Fixes: QTBUG-148523
Change-Id: I98d1d6753d3f4f54dd6b82b34af7c4f496f13c05
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit 1888726c462968b3e0208880beb4839ed7921d01)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/5a8ae372c2ffcf95d0b7b13ecd6d649733d12839
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/gui/image/qicon.cpp                       |  5 +--
 .../image/qiconhighdpi/tst_qiconhighdpi.cpp   | 33 ++++++++++++++++++-
 2 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/src/gui/image/qicon.cpp b/src/gui/image/qicon.cpp
index 67686666934..df26706d880 100644
--- a/src/gui/image/qicon.cpp
+++ b/src/gui/image/qicon.cpp
@@ -317,11 +317,8 @@ QPixmapIconEngineEntry *QPixmapIconEngine::bestMatch(const QSize &size, qreal sc
             imageReader.read(&image);
         }
         if (!image.isNull()) {
-            pe->pixmap.convertFromImage(image);
-            if (!pe->pixmap.isNull()) {
+            if (pe->pixmap.convertFromImage(image))
                 pe->size = pe->pixmap.size();
-                pe->pixmap.setDevicePixelRatio(scale);
-            }
         }
         if (!pe->size.isValid()) {
             removePixmapEntry(pe);
diff --git a/tests/auto/gui/image/qiconhighdpi/tst_qiconhighdpi.cpp b/tests/auto/gui/image/qiconhighdpi/tst_qiconhighdpi.cpp
index 505e1b5ee4f..0c606cd841a 100644
--- a/tests/auto/gui/image/qiconhighdpi/tst_qiconhighdpi.cpp
+++ b/tests/auto/gui/image/qiconhighdpi/tst_qiconhighdpi.cpp
@@ -1,8 +1,9 @@
 // Copyright (C) 2017 The Qt Company Ltd.
 // SPDX-License-Identifier: LicenseRef-Qt-Commercial OR GPL-3.0-only
 
+#include <QIcon>
+#include <QTemporaryFile>
 #include <QTest>
-#include <qicon.h>
 
 class tst_QIconHighDpi : public QObject
 {
@@ -18,6 +19,7 @@ private slots:
     void addPixmap();
     void ninePatch();
     void preferUpscale();
+    void dontModifyDprInCacheEntry();
 };
 
 tst_QIconHighDpi::tst_QIconHighDpi()
@@ -223,6 +225,35 @@ void tst_QIconHighDpi::preferUpscale()
     QCOMPARE(icon.pixmap(QSize(22, 22), 1.25f).devicePixelRatio(), 1.25f);
 }
 
+void tst_QIconHighDpi::dontModifyDprInCacheEntry()
+{
+    std::vector<std::unique_ptr<QTemporaryFile>> files;
+    QIcon icon;
+    const auto createImage = [&](int size, QColor col) {
+        QImage img(QSize(size, size), QImage::Format_ARGB32);
+        img.fill(col);
+        files.emplace_back(std::make_unique<QTemporaryFile>());
+        QTemporaryFile *f = files.back().get();
+        if (!f->open() || !img.save(f, "PNG"))
+            return false;
+        icon.addFile(f->fileName());
+        return true;
+    };
+    QVERIFY(createImage(16, Qt::red));
+    QVERIFY(createImage(20, Qt::green));
+    QVERIFY(createImage(32, Qt::blue));
+
+    auto checkPixmap = [&](qreal dpr) {
+        constexpr QSize reqSize(16, 16);
+        const QPixmap pm = icon.pixmap(icon.actualSize(reqSize), dpr);
+        const QRgb c = pm.toImage().pixel(pm.rect().center());
+        return QColor(c);
+    };
+    QCOMPARE(checkPixmap(1.25), Qt::green);
+    QCOMPARE(checkPixmap(2.00), Qt::blue);
+    QCOMPARE(checkPixmap(1.25), Qt::green);
+}
+
 int main(int argc, char *argv[])
 {
     QGuiApplication app(argc, argv);
